amplipi/streams.py

```diff
@@ -827,26 +827,33 @@ def send_cmd(self, cmd):
       if cmd in self.supported_cmds:
         if cmd == "love": # Pandora specific command
           src_config_folder = f'{utils.get_folder("config")}/srcs/v{self.vsrc}'
-          # Read the file that contains pandora metadata, flip an enum manually as it doesn't do so itself when sending this command
-          with open(f'{src_config_folder}/.config/pianobar/currentSong', 'r+', encoding='utf-8') as file:
-            data = file.read()
-            # This file contains information in a specific format, separated by 3 commas. Example below:
-            # Artist,,,Track,,,Album,,,Album_Art_URL,,,Liked_State,,,Station
-            chunks = data.split(",,,")
-            liked_state = chunks[4].strip()
-
-            # Chunk 5 (index 4) holds the liked state of a song for pandora
-            # Potential values are numbers 0-3, in order: Default, Liked, Disliked, Tired
-            # This code flips it from Default to Liked and vice-versa when you hit the like button, as pianobar only updates metadata when songs start
-            if liked_state == "0":
-              chunks[4] = models.PandoraRating(1)
-            elif liked_state == "1":
-              chunks[4] = models.PandoraRating(0)
-            modified_data = ",,,".join(chunks)
-            file.seek(0)
-            file.truncate()
-            file.write(modified_data)
-            file.flush()
+          file_path = f'{src_config_folder}/.config/pianobar/currentSong'
+          if os.path.exists(file_path):
+            # Read the file that contains pandora metadata, flip an enum manually as it doesn't do so itself when sending this command
+            with open(file_path, 'r+', encoding='utf-8') as file:
+              data = file.read()
+              # This file contains information in a specific format, separated by 3 commas. Example below:
+              # Artist,,,Track,,,Album,,,Album_Art_URL,,,Liked_State,,,Station
+              chunks = data.split(",,,")
+              try:
+                liked_state = models.PandoraRating(int(chunks[4].strip()))
+              except:
+                liked_state = models.PandoraRating.DEFAULT
+
+              # Chunk 5 (index 4) holds the liked state of a song for pandora
+              # Potential values are numbers 0-3, in order: Default, Liked, Disliked, Tired
+              # This code flips it from Default to Liked and vice-versa when you hit the like button, as pianobar only updates metadata when songs start
+              if liked_state == models.PandoraRating.DEFAULT.value:
```

Comment:
We don't support tired and if you dislike a song it immediately skips it so there's no need to change the color of that button

This bit of code is only to flip the thumbs up button from white to green and back to white if it's unliked

Reply:
Looking at how this works again I see a potential race condition in how the pandora rating is stored. Imagine that you thumb up a song right as it is changing to the next song.

1. read currentSong
1. currentSong gets updated to new song
1. we update currentSong with the new Rating ( but the old song is written)
1. AmpliPis interface shows the old song until the pandora station advances to the next song

Reply:
To avoid this race condition we could keep a shadowed current song. The rating code would grab the song from the shadowed current song, if it matches current song it returns the shadowed rating.

Reply:
I think I fixed this in https://github.com/micro-nova/AmpliPi/pull/701/commits/ce61189bfac80557f12b39c6600990d8ac43b1c4, but I feel the change I did there was potentially too large for this scope

I couldn't find a way to easily grab the current metadata state from this spot so I made one. All tests passed when I tried them again.